controllers/workspace_controller.go

```diff
@@ -128,79 +158,177 @@ func (r *WorkspaceReconciler) startJob(jobName string, terraformCmd string, work
+		log.Println("Job Running")
+		if err := r.List(context.Background(), podList, client.InNamespace(workspace.Namespace), client.MatchingLabels(podLabel)); err != nil {
+			return err
+		}
+		for _, pod := range podList.Items {
+			_ = r.SetControllerReference(foundJob, &pod)
+			if err := r.Get(context.Background(), types.NamespacedName{Name: pod.Name, Namespace: pod.Namespace}, &pod); err != nil {
+				return err
+			}
+			if deleteWs {
+				workspacePhase = terraformv1.WorkspaceDeleting
+			} else {
+				workspacePhase = terraformv1.ObjRunning
 			}
-			workspace.Spec.TfState = state
+			r.Recorder.Event(workspace, "Normal", string(workspacePhase), fmt.Sprintf("Job Running - pod/%s created", pod.Name))
+			workspace.PodName = pod.Name
 			if err := r.Update(context.Background(), workspace); err != nil {
-				return ignoreNotFound(err)
+				return err
+			}
+			if err := r.checkPodStatus(&pod, workspace, deleteWs); err != nil {
+				return err
 			}
-			return nil
 		}
-
-		if foundJob.Status.Failed == failedJobs {
-			return fmt.Errorf("job %s failed", foundJob.Name)
+	case foundJob.Status.Active == 0:
+		log.Println("Job Pending")
+		if err := r.updateStatus(workspace, terraformv1.ObjPending, "PendingPodCreation", false); err != nil {
+			return err
 		}
+		r.Recorder.Event(workspace, "Normal", string(workspace.Status.Phase), fmt.Sprintf("Job waiting for pod creation - job/%s", foundJob.Name))
+	}
+	return nil
+}
 
-		log.Printf("Waiting for %s/%s to complete successfully before syncing terraform state\n", foundJob.Namespace, foundJob.Name)
-		time.Sleep(5 * time.Second)
-
-		continue
+// checkPodStatus checks the status of pod created by job and updates workspace status accordingly
+func (r *WorkspaceReconciler) checkPodStatus(pod *corev1.Pod, workspace *terraformv1.Workspace, deleteWs bool) error {
+	var workspacePhase terraformv1.ObjectPhase
+	podKey := types.NamespacedName{Name: pod.Name, Namespace: pod.Namespace}
+	if deleteWs {
+		workspacePhase = terraformv1.WorkspaceDeleting
+	} else {
+		workspacePhase = terraformv1.ObjRunning
+	}
+	for {
+		if err := r.Get(context.Background(), podKey, pod); err != nil {
+			return err
+		}
+		switch pod.Status.Phase {
+		case corev1.PodPending:
+			for _, containerStatus := range pod.Status.ContainerStatuses {
+				if containerStatus.State.Waiting.Reason == "ErrImagePull" || containerStatus.State.Waiting.Reason == "ImagePullBackOff" {
+					if err := r.updateStatus(workspace, terraformv1.ObjFailed, containerStatus.State.Waiting.Reason, false); err != nil {
+						return err
+					}
+					return fmt.Errorf("error in pulling container image - %s", containerStatus.State.Waiting.Reason)
+				} else {
+					if err := r.updateStatus(workspace, workspacePhase, "PodPending", false); err != nil {
+						return err
+					}
+					break
+				}
+			}
+		case corev1.PodSucceeded:
+			err := r.updateStatus(workspace, workspacePhase, "PodSucceeded", false)
+			return err
```

Comment:
So I'm assuming now that expecting an error when updating the status here is intended, since it happens in both the run and workspace controllers. I'd like to understand this more.

Reply:
We are expecting error or nil from the `updateStatus` function. Updated the code accordingly.